lib/util/tmp.js

```diff
@@ -22,16 +23,35 @@ function mktmpdir (cache, opts) {
 }
 
 module.exports.withTmp = withTmp
+
 function withTmp (cache, opts, cb) {
   if (!cb) {
     cb = opts
     opts = null
   }
   opts = TmpOpts(opts)
-  return BB.using(mktmpdir(cache, opts).disposer(rimraf), cb)
+
+  const disposer = (target, result, shouldThrow = false) => {
+    const returnResult = () => {
+      if (shouldThrow) {
+        throw result
+      }
+      return result
+    }
+    return rimraf(target)
+      .then(() => returnResult())
+      .catch(() => returnResult())
+  }
+
+  return mktmpdir(cache, opts).then((target) =>
+    cb(target)
+      .then((result) => disposer(target, result))
+      .catch((err) => disposer(target, err, true))
+  )
```

Comment:
is there a reason why the `disposer` here can't just be described as logical `then` sequences? I would prefer something more straightforward like:

```js
  return mktmpdir(cache, opts).then((target) =>
    cb(target)
      .then((result) => {
        rimraf(target).then(() => result)
      })
      .catch((err) => {
        throw result
      })
  )
```

I'm really clueless about the particularities of **Bluebird** `using/disposer` so feel free to correct me 😊

Reply:
Yeah, it took me a little while to wrap my head around disposers. But here are some docs if you are curious:

http://bluebirdjs.com/docs/api/promise.using.html
http://bluebirdjs.com/docs/api/disposer.html

So in your case, 

```
  return mktmpdir(cache, opts).then((target) =>
    cb(target) // What if cb rejects here
      .then((result) => {
        rimraf(target).then(() => result) // rimraf will never get called
      })
      .catch((err) => {
        throw result
      })
  )
```

Also, the other caveat with disposers is that it should run on `cb` resolve and rejection, if a disposer is successful, it should resolve or reject following the result of `cb` however, if a disposer fails, it should crash the process.

```
In bluebird we've chosen to do the latter because resources are typically scarce. For example, 
if a database connection cannot be disposed of and Bluebird ignores that, the connection pool 
will be quickly depleted and the process will become unusable (all requests that query the 
database will wait forever). Since Bluebird doesn't know how to handle that, the only sensible 
default is to crash the process. That way, rather than getting a useless process that cannot 
fulfill more requests, we can swap the faulty worker with a new one letting the OS clean up the 
resources for us.
```

So technically, my implementation is incorrect, as it does not crash the node process. As it should look something like:

```
  const disposer = (target, result, shouldThrow = false) => {
    const returnResult = () => {
      if (shouldThrow) {
        throw result
      }
      return result
    }
    return rimraf(target)
      .then(() => returnResult())
      .catch((err) => {
        throw err // crash the process
      })
  }
```

Reply:
I see 👍 it's fine to me either way

also it would be nice if we could extract the `using/disposer` logic into a reusable util function

Reply:
Extracted the disposer into a utility function.